### PR TITLE
adding getPermitsRemaining()

### PR DESCRIPTION
**Pull request by snigdhabanda**
A call to getPermitsRemaining() exposes the permits left to be consumed. For example, if you're rate limiting at 2 permits / sec with a permit count of 1L and a timeout period of 0s, you'd have 1 permit remaining after the first acquisition. If you increase the timeout period to 500ms, you'd have 2 permits remaining after the first acquisition. This functionality is salient for monitoring (ie. setting an alert once getPermitsRemaining() falls below a threshold).

### CODE REVIEW

File: misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt [hunk 1]
```diff
@@ -66,13 +66,13 @@ class RateLimiterTest {
...
-    assertThat(rateLimiter.tryAcquire(1L, 499, TimeUnit.MILLISECONDS)).isTrue()
+    assertThat(rateLimiter.tryAcquire(1L, 0, TimeUnit.MILLISECONDS)).isTrue()
     assertThat(ticker.nowMs).isEqualTo(0L)
 
-    assertThat(rateLimiter.tryAcquire(1L, 499, TimeUnit.MILLISECONDS)).isTrue()
+    assertThat(rateLimiter.tryAcquire(1L, 0, TimeUnit.MILLISECONDS)).isTrue()
     assertThat(ticker.nowMs).isEqualTo(0L)
 
-    assertThat(rateLimiter.tryAcquire(1L, 499, TimeUnit.MILLISECONDS)).isFalse()
+    assertThat(rateLimiter.tryAcquire(1L, 0, TimeUnit.MILLISECONDS)).isFalse()
```
Comment (ean5533) [inline misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt:75]: Why these changes?

File: misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt [hunk 2]
```diff
@@ -107,6 +107,45 @@ class RateLimiter private constructor(
...
+      println("windowSizeNs $windowSizeNs")
+      println("sleepNs $sleepNs")
+
+      val timeoutNs = unit.toNanos(timeout)
+
+      // We'd have to sleep too long for this number of permits. Fail fast.
+      if (sleepNs > timeoutNs) return 0L
+
+      // Try to consume permits! If atomicAllocatedUntil changed, we lost a race; loop to try again.
+      if (!atomicAllocatedUntil.compareAndSet(allocatedUntil, newAllocatedUntil)) continue
```
Comment (ean5533) [inline misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:136]: When I saw the method name `getPermitsRemaining` I assumed that it just returned how many permits were remaining, but it looks like this method actually acquires permits (like `tryAcquire`) and then returns how many permits were remaining after acquiring. In other words, it looks like this method is a reimplementation of `tryAcquire`, but it just returns different information.

I think we should do one of two things:

1. Rename this method to something like `tryAcquire` (though we can't use that name exactly because it would conflict), and possibly deprecate the existing method.
2. Trim down this method significantly so that all it does is get how many permits are remaining, and has no side effects at all, including acquiring any permits.

The second option feels simpler. I imagine these methods be used by the caller like:

```
if(!rateLimiter.tryAcquire(...)) {
  throw RuntimeException("too busy, go away")
}

val permitsRemaining = rateLimiter.getPermitsRemaining()
metricRegistry.recordGauge("permits-remaining-percent", permitsRemaining/maxPermits)
```

On one hand, this would be subject to some race conditions, because two threads might take out two permits at the same time, and then both threads would record that there is `X` percent remaining. On the other hand, I don't think it really matters if this number is subject to race conditions; all anyone will typically care about is the `min` value that occurs in a given time period.

tl;dr I recommend chopping this method way down so that all it does is calculate the `permitsRemaining` and returns that.

File: misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt (hunk 1 above)
Comment (snigdhabanda) [reply misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt]: Reverted these

File: misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt [hunk 3]
```diff
@@ -66,53 +79,69 @@ class RateLimiterTest {
...
     assertThat(ticker.nowMs).isEqualTo(0L)
 
     // Exhausted.
+    assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(0L)
     assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isFalse()
     assertThat(ticker.nowMs).isEqualTo(0L)
 
     rateLimiter.permitsPerSecond = 10L
 
+    assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
```
Comment (snigdhabanda) [inline misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt:118]: This assertion is currently failing because we have reached the maxNs in our current bucket. However, since our rate limiting changed, we should be allowing extension into the next bucket of time. I know we decided to not factor in permitCount, but I think this is the case where it's needed.

File: misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt [hunk 4]
```diff
@@ -66,53 +79,69 @@ class RateLimiterTest {
...
     assertThat(rateLimiter.tryAcquire(1L, 100, TimeUnit.MILLISECONDS)).isTrue()
     assertThat(ticker.nowMs).isEqualTo(100L)
+    assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(0L)
   }
 
   @Test
   fun `rate limit decreases`() {
     rateLimiter.permitsPerSecond = 10L
 
+    assertThat(rateLimiter.getPermitsRemaining(TimeUnit.MILLISECONDS, 100)).isEqualTo(1L)
```
Comment (snigdhabanda) [inline misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt:128]: Similar to above, this assertion fails. I would think here that only permit is allowed, but the behavior is that 10 permits are actually allowed. I think taking permitCount into account would once again prevent this.

File: misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt (hunk 3 above)
Comment (ean5533) [reply misk-core/src/test/kotlin/misk/sampling/RateLimiterTest.kt:118]: The `permitCount` parameter from `tryAcquire` represents how many permits the caller is asking for. This should not have any impact on the calculation of how many permits are available.

(disussed in another forum; we're changing the calculation to be simpler, but still not using `permitCount`)

File: misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt [hunk 5]
```diff
@@ -107,6 +108,33 @@ class RateLimiter private constructor(
     }
   }
 
+  /**
+   * Returns the permits remaining, given a time unit and timeout, after tryAcquire() has been invoked.
```
Comment (ean5533) [inline misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:112]: I think this comment is a bit misleading. There's no requirement for this method to be called before or after `tryAcquire` has been invoked; it can be called any time, even if `tryAcquire` is never invoked.

Perhaps something like this?

> Returns the maximum number of permits that could have been acquired by a call to [tryAcquire], assuming the caller passed the same [timeout] and [unit].

**Review by ean5533 — APPROVED** [review]
One nit about the doc comment, but otherwise this LGTM.

I recommend getting a review from @isafan or @mightyguava if possible since they wrote/reviewed the original implementation. Otherwise, someone like @adrw or @swankjesse would be good.

File: misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt [hunk 6]
```diff
@@ -81,17 +81,18 @@ class RateLimiter private constructor(
...
       if (permitCount > maxRequestSize) return null
-
       val now = ticker.read()
+
       val timeoutNs = unit.toNanos(timeout)
 
       // If this acquire succeeds, this is the time we're consuming permits through.
-      val newAllocatedUntil = maxOf(allocatedUntil, now) +
-        permitCount.permitsToNanos(permitsPerSecond)
+      val newAllocatedUntil =
```
Comment (adrw) [inline misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:94]: nit: I prefer the previous line break

File: misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt [hunk 7]
```diff
@@ -107,6 +108,35 @@ class RateLimiter private constructor(
     }
   }
 
+  /**
+   * Returns the maximum number of permits that could have
+   * been acquired by a call to [tryAcquire], assuming the caller
+   * passed the same [timeout] and [unit].
+   *
+   * @return the permits remaining.
```
Comment (adrw) [inline misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:116]: nit: I think this `@return` line is redundant and implied by the function name and could be removed, thanks for the above description though!

File: misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt [hunk 8]
```diff
@@ -107,6 +108,35 @@ class RateLimiter private constructor(
...
+   */
+  fun getPermitsRemaining(
+    unit: TimeUnit,
+    timeout: Long
+  ): Long {
+    val allocatedUntil = atomicAllocatedUntil.get()
+    println("getPermitsRemaining.allocatedUntil: $allocatedUntil")
```
Comment (adrw) [inline misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:123]: this should be removed before merge

Comment (adrw) [reply misk-core/src/main/kotlin/misk/sampling/RateLimiter.kt:123]: (I'm assuming it is there solely for debugging during development and not intended to land in master)

**Review by adrw — APPROVED** [review]
